expr/shaper.go

```diff
@@ -217,154 +217,163 @@ func (s *step) buildRecord(in zcode.Bytes, b *zcode.Builder) {
+}
+
+// NewShaper returns a shaper that will shape the result of expr
+// to the type returned by typExpr.
+func NewShaper(zctx *zson.Context, expr, typExpr Evaluator, tf ShaperTransform) *Shaper {
+	return &Shaper{
+		zctx:       zctx,
+		typExpr:    typExpr,
+		expr:       expr,
+		shapers:    make(map[zng.Type]*ConstShaper),
+		transforms: tf,
 	}
-	// XXX This needs to be done at runtime because the type expression can
-	// depend on typenames defined in the data.  See issue #2383.
-	typVal, err := typExpr.Eval(nil)
+}
+
+func (s *Shaper) Eval(rec *zng.Record) (zng.Value, error) {
+	typVal, err := s.typExpr.Eval(rec)
 	if err != nil {
-		return nil, err
+		return zng.Value{}, err
 	}
 	if typVal.Type != zng.TypeType {
-		return nil, fmt.Errorf("shaping functions (crop, fill, cast, order) require a type value as second parameter")
-	}
-	s, err := zng.DecodeString(typVal.Bytes)
-	if err != nil {
-		return nil, err
+		return zng.NewErrorf("shaper function type argument is not a type"), nil
 	}
-	shapeType, err := zctx.Context.LookupByName(s)
-	if err != nil {
-		return nil, fmt.Errorf("shaper could not parse type value literal: %s", err)
+	shapeTo := s.zctx.FromTypeValue(typVal)
+	if shapeTo == nil {
+		return zng.NewErrorf("shaper encountered unknown type value: %q", string(typVal.Bytes)), nil
 	}
-
-	if _, ok := zng.AliasOf(shapeType).(*zng.TypeRecord); !ok {
-		return nil, fmt.Errorf("shaping functions (crop, fill, cast, order) require a record type value as second parameter (got %T %T)", shapeType, zng.AliasOf(shapeType))
+	shaper, ok := s.shapers[shapeTo]
+	if !ok {
+		if _, ok := zng.AliasOf(shapeTo).(*zng.TypeRecord); !ok {
+			return zng.NewErrorf("shaper function type argument is not a record type: %q", shapeTo.ZSON()), nil
+		}
+		shaper = NewConstShaper(s.zctx, s.expr, shapeTo, s.transforms)
+		s.shapers[shapeTo] = shaper
 	}
-	return NewShaper(zctx, fieldExpr, shapeType, tf)
+	return shaper.Eval(rec)
 }
 
-func (s *Shaper) Apply(in *zng.Record) (*zng.Record, error) {
+func (s *ConstShaper) Apply(in *zng.Record) (*zng.Record, error) {
 	v, err := s.Eval(in)
 	if err != nil {
 		return nil, err
 	}
 	return zng.NewRecord(v.Type.(*zng.TypeRecord), v.Bytes), nil
 }
 
-func (s *Shaper) Eval(in *zng.Record) (zng.Value, error) {
-	inVal, err := s.fieldExpr.Eval(in)
+func (c *ConstShaper) Eval(in *zng.Record) (zng.Value, error) {
+	inVal, err := c.expr.Eval(in)
 	if err != nil {
 		return zng.Value{}, err
 	}
 	inType, ok := inVal.Type.(*zng.TypeRecord)
 	if !ok {
 		return inVal, nil
 	}
-	if _, ok := s.shapeSpecs[in.Type.ID()]; !ok {
-		spec, err := s.createShapeSpec(inType)
+	id := in.Type.ID()
+	s, ok := c.shapers[id]
+	if !ok {
+		s, err = createShaper(c.zctx, c.transforms, c.shapeTo, inType)
 		if err != nil {
 			return zng.Value{}, err
 		}
-		s.shapeSpecs[in.Type.ID()] = spec
+		c.shapers[id] = s
 	}
-	spec := s.shapeSpecs[in.Type.ID()]
-	if spec.typ.ID() == in.Type.ID() {
+	if s.typ.ID() == in.Type.ID() {
```

Comment:
```suggestion
	if s.typ.ID() == id {
```

Reply:
I realize this should return a value with s.typ.  I'll add a test for it too.